### PR TITLE
FE: Add retry logic to router for get blob

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
This PR adds retry logic for GetBlobOperation in router.

Why are we doing this:
Router might fail client's get requests due to temporary network error, or server being overloaded. If frontend returns 503 right away, client would would probably retry the entire blob. However, if this is a composite blob made of hundreds of chunks, then client resending get request would end up wasting lots of resources to refetch those already fetched blob in router.

This PR adds retry logic to prevent resource from wasting. It will retry on GetBlobOperation. But there is a limit to which an GetBlobOperation can retry. This limit is based on both count and duration. We can configure the maximum number of retries and the longest duration for retries. A retry can't exceed either limit.

### CODE REVIEW

**Review by mikeyang923 — APPROVED** [review]

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1436]: Do we have issue here? Looks like either way, chunkException will be reset before the PutBlob retry, right?
Somehow for PutBlob, I like the current implementation more since we consolidate all the initial setting all in  prepareForSending.
But your call.

File: ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java
Comment (JingQianCloud) [inline ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java:1178]: Can we add unit test for composite blob and test retry on multiple chunks?

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:813]: If for some reason, the servers returns the server unavailable very quickly and there are not many replicas online for this partition, our frontend retry may be too stressful to these remaining replicas.

I was thinking about some simple backoff.
But we cannot do 
"sleep(10ms * failedAttempts);"
It'll block the RequestResponseHandlerThread which is shared by different Operations.

Do we have some other way for some simple backoff? Probably record the time and when it reaches that time + delay time, start the next retry. Not sure if the code will be too complicated.

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:663]: lets get rid of the Defaults. they are not used anywhere and are jsut confusing

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java
Comment (snalli) [inline ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:675]: 100 seems high

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (snalli) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1042]: Nice ! We are retrying with num failed attempts and a timeout

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1038]: the method processServerError should do it. Basically, the RouterException for Retry_After_Backoff is UnexpectedInternalError for short.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:1436]: right, it's just prepareForSending should not reset the chunkException. The exception should be reset in the checkAndMaybeComplete method. It's easier to understand that way.

File: ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java
Comment (justinlin-linkedin) [reply ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java:1178]: will do

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:813]: kinda hard to do this, we have to add another configuration for backoff, the time to sleep etc... we will be introducing too many new configurations.

**Review by JingQianCloud — APPROVED** [review]

**Comment by snalli** [comment]
can u pls address my comments before merging ?

**Comment by snalli** [comment]
can u pls address my comments before merging ?

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java
Comment (ankagrawal) [reply ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:675]: +1 Maybe 3 or 5 is good enough?

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (ankagrawal) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1038]: Should all types of `UnexepectedInternalError` lead to retry? Like if somehow the server returns `BlobDeleted` for data chunks, then we should not retry them here.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java
Comment (ankagrawal) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1044]: nit: Maybe calculate `TimeUnit.SECONDS.toMillis(routerConfig.routerGetBlobRetryLimitInSec)` once in the constructor.

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:673]: this is the maximum number, I don't see we are setting the value to this number. But I don't want to limit it under 100.

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java]: done

**Pull request merged**
